hw/ip/prim/doc/prim_flash.md

```diff
@@ -0,0 +1,104 @@
+TestModeWidth  | int    | The number of test modes for a bank of flash
+
+
+## Signal Interfaces
+
+### Overall Interface Signals
+Name                 | In/Out | Description
+---------------------|--------|---------------------------------
+clk                  | input  | Clock input
+rst_n                | input  | Reset input
+flash_req            | input  | Inputs from flash protocol and physical controllers
+flash_rsp            | output | Outputs to flash protocol and physical controllers
+prog_type_avail      | output | Avaialble program types in this flash wrapper: Currently there are only two types, program normal and program repair
+init_busy            | output | The flash wrapper is undergoing initialization
+tck                  | input  | jtag tck
+tdi                  | input  | jtag tdi
+tms                  | input  | jtag tms
+tdo                  | output | jtag tdo
+scanmode             | input  | dft scanmode input
+scan_rst_n           | input  | dft scanmode reset
+flash_power_ready_h  | inout  | flash power is ready (high voltage connection)
+flash_power_down_h   | inout  | flash wrapper is powering down (high voltage connection)
+flash_test_mode_a    | inout  | flash test mode values (analog connection)
+flash_test_voltage_h | inout  | flash test mode voltage (high voltage connection)
+
+
+### Flash Request/Response Signals
+
+Name         | In/Out | Description
+-------------|--------|---------------------------------
+rd           | input  | read request
+prog         | input  | program request
+prog_last    | input  | last program beat
+prog_type    | input  | type of program requested: currently there are only two types, program normal and program repair
+pg_erase     | input  | page erase request
+bk_erase     | output | bank erase request
+addr         | input  | requested transaction address
+part         | input  | requested transaction partition
+info_sel     | input  | if requested transaction is information partition, the type of information partition accessed
+he           | output | high endurance enable for requested address
+prog_data    | input  | program data
+ack          | output | transction acknowledgement
+rd_data      | output | transaction done
+
+
+# Theory of Operations
+
+## Transactions
+
+Transactions into the flash wrapper follow a req / ack / done format.
+A request is issued by raising one of `rd`, `prog`, `pg_erase` or `bk_erase` to 1.
+When the flash wrapper accepts the transaction, `ack` is returned.
+When the transaction fully completes, a `done` is sent returned as well.
+
+Depending on the type of transaction, there may be a significant gap between `ack` and `done`.
+For example, a read may have only 1 or 2 cycles between transaction acknowledgement and transaction complete.
+Whereas a program or erase may have a gap extending up to uS or even mS.
+
+It is the flash wrapper's decision on how many outstanding transaction to accept.
+The following diagram is one such example:
+
```

Comment:
missing diagram?

Reply:
ah yes this one i'm going to steal Nuvoton's :)
Haven't gotten the source yet.